dgraphtest/cluster.go

```diff
@@ -0,0 +1,301 @@
+		zo.coname = fmt.Sprintf(zeroNameFmt, c.conf.prefix, zo.id)
+		zo.loname = fmt.Sprintf(zeroLNameFmt, zo.id)
+		cid, err := c.createContainer(zo)
+		if err != nil {
+			return err
+		}
+		zo.coid = cid
+		c.zeros = append(c.zeros, zo)
+	}
+
+	for i := 0; i < c.conf.numAlphas; i++ {
+		aa := &alpha{id: i}
+		aa.coname = fmt.Sprintf(alphaNameFmt, c.conf.prefix, aa.id)
+		aa.loname = fmt.Sprintf(alphaLNameFmt, aa.id)
+		cid, err := c.createContainer(aa)
+		if err != nil {
+			return err
+		}
+		aa.coid = cid
+		c.alphas = append(c.alphas, aa)
+	}
+
+	return nil
+}
+
+func (c *Cluster) createNetwork() error {
+	c.net.name = c.conf.prefix + "-net"
+	opts := types.NetworkCreate{
+		Driver: "bridge",
+		IPAM:   &network.IPAM{Driver: "default"},
+	}
+
+	network, err := c.dcli.NetworkCreate(context.Background(), c.net.name, opts)
+	if err != nil {
+		return errors.Wrap(err, "error creating network")
+	}
+	c.net.id = network.ID
+
+	return nil
+}
+
+func (c *Cluster) Start() error {
+	c.log("starting cluster with prefix [%v]", c.conf.prefix)
+	for i := 0; i < c.conf.numZeros; i++ {
+		if err := c.StartZero(i); err != nil {
+			return err
+		}
+	}
+	for i := 0; i < c.conf.numAlphas; i++ {
+		if err := c.StartAlpha(i); err != nil {
+			return err
+		}
+	}
+
+	if err := c.healthCheck(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Cluster) StartZero(id int) error {
+	if id >= c.conf.numZeros {
+		return fmt.Errorf("invalid id of zero: %v", id)
+	}
+	return c.startContainer(c.zeros[id])
+}
+
+func (c *Cluster) StartAlpha(id int) error {
+	if id >= c.conf.numAlphas {
+		return fmt.Errorf("invalid id of alpha: %v", id)
+	}
+	return c.startContainer(c.alphas[id])
+}
+
+func (c *Cluster) startContainer(dc dnode) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	if err := c.dcli.ContainerStart(ctx, dc.cid(), types.ContainerStartOptions{}); err != nil {
+		return errors.Wrapf(err, "error starting container [%v]", dc.cname())
+	}
+	return nil
+}
+
+func (c *Cluster) Stop() error {
+	c.log("stopping cluster with prefix [%v]", c.conf.prefix)
+	for i := range c.alphas {
+		if err := c.StopAlpha(i); err != nil {
+			return err
+		}
+	}
+	for i := range c.zeros {
+		if err := c.StopZero(i); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Cluster) StopZero(id int) error {
+	if id >= c.conf.numZeros {
+		return fmt.Errorf("invalid id of zero: %v", id)
+	}
+	return c.stopContainer(c.zeros[id])
+}
+
+func (c *Cluster) StopAlpha(id int) error {
+	if id >= c.conf.numAlphas {
+		return fmt.Errorf("invalid id of alpha: %v", id)
+	}
+	return c.stopContainer(c.alphas[id])
+}
+
+func (c *Cluster) stopContainer(dc dnode) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	if err := c.dcli.ContainerStop(ctx, dc.cid(), &stopTimeout); err != nil {
+		return errors.Wrapf(err, "error stopping container [%v]", dc.cname())
+	}
+	return nil
+}
+
```

Comment:
Probably not a part of this diff, but can we also include options for collecting logs here?

Reply:
yes, for sure